### DISCUSSION TITLE
Extend `with` to include immutable indexed collections

### DISCUSSION THREAD

**Discussion by cgranade** [dotnet/csharplang#3474] [Language Ideas]
I was recently very excited to see the new [`with` expresion](https://devblogs.microsoft.com/dotnet/welcome-to-c-9-0/#with-expressions) for working with immutable record types, and wanted to suggest extending this new feature to include set-and-replace operations on immutable list and dictionary items as well. For example:

```C#
var list = (new [] { 0, 1, 2 }).ToImmutableList()
           with { [1] = 10 };
var builder = ImmutableDictionary.CreateBuilder<string, int>();
builder.Add("foo", 42);
builder.Add("bar", 12);
var dict = builder.ToImmutable() with { ["bar"] = 314 };
```

We use a similar syntax in Q# for immutable array set-and-replace operations (e.g.: `ConstantArray(4, "I") w/ 1 <- "X"` is the array `["I", "X", "I", "I"]`) and have found it to be a really useful feature.

**Comment by svick** [comment]
Is being able to set a specific value actually useful, if you can't use a similar syntax for even such basic operations as adding or removing items? That doesn't fell very consistent to me.

**Comment by spydacarnage** [comment]
You can already use .SetItem() to "modify" an item in an Immutable collection
